### PR TITLE
mapl: add 2.40, deprecate old versions

### PR DESCRIPTION
**Pull request by mathomp4**
This PR adds the latest MAPL 2.40 versions. It also does a couple other things:

1. We deprecate MAPL 2.12 and older versions in the hope that they can be removed "soon" (aka the next release, see https://github.com/spack/spack/pull/39227#issuecomment-1663103136)
2. Make the `flap` variant only for MAPL 2.39 and lower since we removed FLAP support from MAPL in 2.40.
3. Changed some variant defaults. Per a suggestion of @AlexanderRichert-NOAA, the variants now reflect how CMake `options` default. So that means `+fargparse +pflogger +extdata2g +shared` and all the rest off.
4. Removed the `pnetcdf` variant. MAPL has no support for pnetcdf so this is sort of useless code. All it does is maybe add a build of `parallel-netcdf` for no good reason.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mapl/package.py [hunk 1]
```diff
@@ -138,27 +182,38 @@ class Mapl(CMakePackage):
     depends_on("netcdf-c")
     depends_on("netcdf-fortran")
     depends_on("parallel-netcdf", when="+pnetcdf")
-    depends_on("esmf@8.4:", when="@2.34:")
-    depends_on("esmf@8.3:", when="@2.22:")
+    depends_on("esmf@8.5:", when="@2.40:")
+    depends_on("esmf@8.4.1:8.4", when="@2.34:2.39")
+    depends_on("esmf@8.3.0:8.3", when="@2.22:2.33")
```
Comment (AlexanderRichert-NOAA) [inline var/spack/repos/builtin/packages/mapl/package.py]: I think you should be able to do, e.g., `esmf@8.4` and that will pick up anything in the 8.4 range.

**Comment by AlexanderRichert-NOAA** [comment]
As far as spack-stack we're fine either way in terms of the variant defaults; since we have a heads up we'll keep it in mind when we update the recipe next and make sure are settings are all good. Since you're the developer I'll defer to you on what the defaults should be. My two cents is that I usually try to have the variants match with cmake defaults unless there's some particular reason not to, but, that's just me.

**Comment by climbfuji** [comment]
@mathomp4 If you don't have any pnetcdf in your code, why does that variant exist? Should it be removed altogether?

**Comment by mathomp4** [comment]
> As far as spack-stack we're fine either way in terms of the variant defaults; since we have a heads up we'll keep it in mind when we update the recipe next and make sure are settings are all good. Since you're the developer I'll defer to you on what the defaults should be. My two cents is that I usually try to have the variants match with cmake defaults unless there's some particular reason not to, but, that's just me.

@AlexanderRichert-NOAA That actually does make sense! Let me do that.


> @mathomp4 If you don't have any pnetcdf in your code, why does that variant exist? Should it be removed altogether?

Maybe? That variant was there in the original package I started with (from hpc-stack maybe? Not sure.) I'm fearful of removing it since, well, maybe someone has some clever way of using it in MAPL?

@tldahlgren Can one "deprecate" a variant? I mean, I'd love to do something to say "We don't know what this variant does, but we want to know if someone tries to build with it."

**Review by tldahlgren — DISMISSED** [review]
Confirmed the 15 version sha256.  Deferring merge to give time for you all to work out the rest.

**Comment by mathomp4** [comment]
@tldahlgren Do you know if there is an "official" way to deprecate/remove variants? I'd like to throw a warning for a while if anyone tries to use the pnetcdf variant as, frankly, I'm not sure it does anything. 

Is the best I can think of is to do some sort of:
```python
if pnetcdf in self.spec:
    print("The pnetcdf variant is untested and deprecated. Contact maintainers if you use this")
```
??

**Comment by climbfuji** [comment]
@mathomp4 If you know that there is no code in mapl that uses pnetcdf, then I'd be ok taking a pragmatic approach and just removing the variant.

**Comment by mathomp4** [comment]
> @mathomp4 If you know that there is no code in mapl that uses pnetcdf, then I'd be ok taking a pragmatic approach and just removing the variant.

@climbfuji Yeah. I just consulted with @bena-nasa (our MAPL netCDF Guru) and he's pretty sure pnetcdf has never been a part of MAPL, only parallel-HDF5-style netcdf4. So, I think I'll just remove it.

ETA: Removed.

File: var/spack/repos/builtin/packages/mapl/package.py (hunk 1 above)
Comment (climbfuji) [reply var/spack/repos/builtin/packages/mapl/package.py]: @mathomp4 Can you look at Alex's suggestion, please? I think that's the last thing.

Comment (mathomp4) [reply var/spack/repos/builtin/packages/mapl/package.py]: Well huh. I *know* I did that! I guess I never pushed. One moment...

Comment (mathomp4) [reply var/spack/repos/builtin/packages/mapl/package.py]: Tested and pushed.

**Comment by tldahlgren** [comment]
> @tldahlgren Can one "deprecate" a variant? I mean, I'd love to do something to say "We don't know what this variant does, but we want to know if someone tries to build with it."

Well, to my knowledge, you cannot directly deprecate a variant.  

You CAN make a variant conditional (https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-variants) by adding the `when` argument.

If you want to output a message (and disallow the variant's use), you could use a conflict (see https://spack.readthedocs.io/en/latest/packaging_guide.html#common-when-constraints) with the `msg` argument.

**Review by AlexanderRichert-NOAA — APPROVED** [review]

**Review by climbfuji — APPROVED** [review]

**Review by tldahlgren — APPROVED** [review]

**Comment by climbfuji** [comment]
@mathomp4 Do you want to cherry-pick this for our fork or should I ?

**Comment by mathomp4** [comment]
> @mathomp4 Do you want to cherry-pick this for our fork or should I ?

You might be better to do so. The last time I tried cherry-pick with git I screwed up royally. It's like my brain said "cherry-pick is not recommended!" 😄

**Pull request merged**